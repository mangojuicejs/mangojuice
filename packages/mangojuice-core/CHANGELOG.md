# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.3"></a>
## 0.1.3 (2017-07-12)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))
* **core:** recursive nesting works now ([c462be5](https://github.com/mangojuicejs/mangojuice/commit/c462be5))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)
