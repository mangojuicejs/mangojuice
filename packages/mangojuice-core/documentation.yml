toc:
  - name: Basic
    description: |
      The most common APIs that you will work with most of the time
    children:
      - Block
      - LogicBase
      - cmd
      - run
      - mount
      - task
      - cancel
      - child
      - depends
  - name: Extras
    description: |
      Other pieces of MangoJuice
    children:
      - Process
      - Command
      - TaskMeta
      - DefaultLogger
      - ComputedField
      - Mounter
      - defineCommand
      - ensureCommand
      - decorateLogic
      - handleBefore
      - handleAfter
      - handle
      - call
      - logicOf
      - observe
      - procOf
      - bind
      - hydrate
      - delay
