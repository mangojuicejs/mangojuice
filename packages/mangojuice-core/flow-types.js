// @flow
type CommandType<M> = {
  is: (...args: any) => bool,
  model: M,
};

type TaskObject = {
  success: (cmd: any) => TaskObject,
  fail: (cmd: any) => TaskObject,
  args: (...args: Array<any>) => TaskObject,
  engine: (exec: any) => TaskObject
};

type DependsType = {
  compute: (Function) => DependsType
};

type LogicConfig<MetaType> = {
  initCommands?: any,
  meta?: MetaType
};

type ConfiguredLogic = {
};

type BlockType = {
  createModel: (...args: any) => Object,
  Logic: Class<*>,
  View?: any
};

declare module "mangojuice-core" {
  declare export function cmd(): any;

  declare export function debounce(): any;

  declare export function throttle(): any;

  declare export function task(fn: Function): TaskObject;

  declare export function delay(ms: number): Promise<void>;

  declare export function cancel(cmd: any): any;

  declare export function logicOf<T>(model: Object, clazz?: Class<T>): T;

  declare export function child<T>(logicClass: Class<{ +config?: (...args: T) => any }>, ...args: T): ConfiguredLogic;

  declare export function depends(...args: any): DependsType;

  declare export function observe(model: Object, destroyeded: Promise<any>, handler: Function): void;

  declare export function run(block: BlockType, opts?: Object): any;

  declare export function bind(block: BlockType, opts?: Object): any;

  declare export function hydrate(block: BlockType, model: Object): any;

  declare export var utils: any;

  declare export class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
    +model: ModelType;
    +shared: SharedType;
    +meta: MetaType;

    hub(cmd: CommandType<*>): ?Promise<any>;

    hubBefore(cmd: CommandType<*>): ?Promise<any>;

    hubAfter(cmd: CommandType<*>): ?Promise<any>;

    port(exec: (cmd: any) => Promise<any>, destroyed: Promise<any>): ?Promise<any>;

    children(): {[k: $Enum<ModelType>]: Class<*> | ConfiguredLogic};

    computed(): {[k: $Enum<ModelType>]: (() => any) | DependsType};

    config(...args: any): LogicConfig<MetaType>;
  }

  declare export default {|
    +LogicBase: typeof LogicBase,
    +cmd: typeof cmd,
    +debounce: typeof debounce,
    +throttle: typeof throttle,
    +task: typeof task,
    +delay: typeof delay,
    +cancel: typeof cancel,
    +logicOf: typeof logicOf,
    +child: typeof child,
    +depends: typeof depends,
    +observe: typeof observe,
    +run: typeof run,
    +bind: typeof bind,
    +hydrate: typeof hydrate,
    +utils: typeof utils
  |};
}
