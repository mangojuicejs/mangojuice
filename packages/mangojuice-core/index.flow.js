// @flow

type NestType<A, T, Y, U> = {
  +logic: A,
  args: (c1?: T, c2?: Y, c3?: U) => NestType<A, T, Y, U>,
  handler: () => NestType<A, T, Y, U>,
  singleton: (?Boolean) => NestType<A, T, Y, U>
};

type DependsType = {
  compute: (Function) => DependsType
};

interface ConfigFunction<MetaType, T, Y, U> {
  +config?: (c1?: T, c2?: Y, c3?: U) => ConfigType<MetaType>;
};

type ConfigType<M> = {
  initCommands?: any,
  meta?: M
};

declare class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
  +model: ModelType;
  +shared: SharedType;
  +meta: MetaType;
  +destroy: Promise<any>;

  port(): ?Promise<any>;

  children(): {[key: $Enum<ModelType>]: NestType<*, *, *, *>};

  computed(): {[k: $Enum<ModelType>]: () => any | DependsType};

  exec(): Promise<any>;

  depends(...deps: Array<Object>): DependsType;

  nest<T, Y, U, L: ConfigFunction<MetaType, T, Y, U>&LogicBase<*,SharedType,*>>(logic: L): NestType<L, T, Y, U>;
}
