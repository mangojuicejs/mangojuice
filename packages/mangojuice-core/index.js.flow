// @flow

// Cmd
//
declare export var Cmd: {
  update: any,
  batch: any,
  execLatest: any,
  execEvery: any,
  debounce: any,
  throttle: any
};

// Task
//
type TaskObject = {
  success: (cmd: any) => TaskObject,
  fail: (cmd: any) => TaskObject
};
declare export var Task: {
  create: (Function) => TaskObject;
};


// Run
//
declare export var Run: any;


// LogicBase
//
type NestType<A, T, Y, U> = {
  +logic: A,
  args: (c1?: T, c2?: Y, c3?: U) => NestType<A, T, Y, U>,
  handler: () => NestType<A, T, Y, U>,
  singleton: (?Boolean) => NestType<A, T, Y, U>
};

type DependsType = {
  compute: (Function) => DependsType
};

interface ConfigFunction<MetaType, T, Y, U> {
  +config?: (c1?: T, c2?: Y, c3?: U) => ConfigType<MetaType>;
};

type ConfigType<M> = {
  initCommands?: any,
  meta?: M
};

declare export class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
  +model: ModelType;
  +shared: SharedType;
  +meta: MetaType;
  +destroy: Promise<any>;

  port(): ?Promise<any>;

  children(): {[key: $Enum<ModelType>]: NestType<*, *, *, *>};

  computed(): {[k: $Enum<ModelType>]: () => any | DependsType};

  exec(): Promise<any>;

  depends(...deps: Array<Object>): DependsType;

  nest<T, Y, U, L: ConfigFunction<MetaType, T, Y, U>&LogicBase<*,SharedType,*>>(logic: L): NestType<L, T, Y, U>;
}
