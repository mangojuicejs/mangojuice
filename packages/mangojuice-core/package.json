{
  "name": "mangojuice-core",
  "version": "0.1.6",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs"
  }
}
