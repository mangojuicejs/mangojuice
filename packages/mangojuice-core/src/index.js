import Process from './classes/Process';
import LogicBase from './classes/LogicBase';
import DefaultLogger from './classes/DefaultLogger';
import child from './core/child';
import msg from './core/msg';
import update from './core/update';
import depends from './core/depends';
import context from './core/context';
import observe from './core/observe';
import debounce from './core/debounce';
import throttle from './core/throttle';
import logicOf from './core/logicOf';
import procOf from './core/procOf';
import delay from './core/task/delay';
import task from './core/task/task';
import run from './core/run/run';
import bind from './core/run/bind';
import mount from './core/run/mount';
import * as utils from './core/utils';
import * as config from './config';


export {
  // Classes
  Process,
  LogicBase,
  DefaultLogger,
  // Core
  child,
  msg,
  depends,
  context
  observe,
  update,
  debounce,
  throttle,
  logicOf,
  procOf,
  delay,
  task,
  run,
  bind,
  mount,
  utils,
  config
};
