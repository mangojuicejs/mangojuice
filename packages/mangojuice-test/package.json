{
  "name": "mangojuice-test",
  "version": "2.0.0-y.0",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs.Test",
    "bundledDependencies": [],
    "moduleGlobals": {
      "mangojuice": "Mjs"
    }
  },
  "dependencies": {
    "mangojuice-core": "2.0.0-y.0"
  }
}
