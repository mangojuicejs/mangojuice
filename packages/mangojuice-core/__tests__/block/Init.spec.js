import { update, depends, child, task, delay, event, context } from 'mangojuice-core';
import { runWithTracking } from 'mangojuice-test';

describe('Init commands execution', () => {
  const AsyncTaskDelayed = function() {
    return this.call(delay, 50);
  };
  const SharedBlock = {
    Events: {
      UpdateShared: event()
    },
    Logic: class SharedBlockLogic {
      prepare() {
        return this.FromInitOneCmd();
      }
      hub(event) {
        if (event.is(SharedBlock.Events.UpdateShared)) {
          return { SharedUpdated: { ...event } };
        }
      }
      FromPortAsync() {
        return task(AsyncTaskDelayed).success(this.FromPortAsync_Success);
      }
      FromInitOneCmd() {
        return { FromInitOneCmd: true };
      }
      FromPortAsync_Success() {
        return { FromPortAsync_Success: true };
      }
    }
  };
  const BlockB = {
    Events: {
      SomeUpdate: event()
    },
    Logic: class BlockBLogic {
      prepare() {
        return [
          { sharedDump: JSON.stringify(this.shared) },
          this.FromInitOneCmd
        ];
      }
      hub(event) {
        if (event.is(BlockB.Events.SomeUpdate)) {
          return { SomeEventUpdated: { ...event } };
        }
      }
      FromInitOneCmd() {
        return [
          { FromInitOneCmd: true },
          SharedBlock.Events.UpdateShared(3,2,1)
        ];
      }
    }
  };
  const BlockA = {
    Logic: class BlockALogic {
      prepare() {
        return [
          context({
            some: child(SharedBlock.Logic)
          }),
          {
            b_1: child(BlockB.Logic),
            b_2: child(BlockB.Logic)
          },
          this.FromInitOneCmd,
          this.FromInitTwoCmd(1, 2, 3)
        ];
      }
      hub(event, fromChildren) {
        if (event.is(SharedBlock.Events.UpdateShared) && fromChildren) {
          return { ReactOnUpdateShared: { ...event } };
        }
      }
      FromPortCmd() {
        return [ this.FromPortCmd_1, this.FromPortCmd_2() ];
      }
      FromPortAsync() {
        return task(AsyncTaskDelayed)
          .success(this.FromPortAsync_Success);
      }
      FromPortCmd_1() { return { FromPortCmd_1: true }; }
      FromPortCmd_2() { return { FromPortCmd_2: true }; }
      FromPortAsync_Success() { return { FromPortAsync_Success: true } }
      FromInitOneCmd() { return { FromInitOneCmd: true } }
      FromInitTwoCmd() { return { FromInitTwoCmd: true } }
      FromSubCmd() { return [
        { b_1: update(BlockB.Logic, BlockB.Events.SomeUpdate(1,2,3)) },
        { FromSubCmd: true }
      ] }
      HandleB_1() { return { HandleB_1: true } }
      HandleB_11() { return { HandleB_11: true } }
      HandleB_2() { return { HandleB_2: true } }
      HandleB_22() { return { HandleB_22: true } }
    }
  };

  it('should execute init commands in correct order and finish', async () => {
    const { app, commandNames, errors } = await runWithTracking({
      app: BlockA
    });

    expect(errors).toMatchSnapshot();
    expect(app.model).toMatchSnapshot();
    expect(commandNames).toMatchSnapshot();
  });
});
