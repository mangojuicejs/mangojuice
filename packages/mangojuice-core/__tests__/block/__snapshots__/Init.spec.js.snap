// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init commands execution should execute init commands in correct order and finish 1`] = `Array []`;

exports[`Init commands execution should execute init commands in correct order and finish 2`] = `
Object {
  "FromInitOneCmd": true,
  "FromInitTwoCmd": true,
  "FromPortAsync_Success": true,
  "FromPortCmd_1": true,
  "FromPortCmd_2": true,
  "FromSubCmd": true,
  "ReactOnUpdateShared": Object {
    "0": 3,
    "1": 2,
    "2": 1,
  },
  "b_1": Object {
    "FromInitOneCmd": true,
    "SomeEventUpdated": Object {
      "0": 1,
      "1": 2,
      "2": 3,
    },
    "sharedDump": "{\\"some\\":{\\"FromInitOneCmd\\":true}}",
  },
  "b_2": Object {
    "FromInitOneCmd": true,
    "sharedDump": "{\\"some\\":{\\"FromInitOneCmd\\":true,\\"SharedUpdated\\":{\\"0\\":3,\\"1\\":2,\\"2\\":1}}}",
  },
}
`;

exports[`Init commands execution should execute init commands in correct order and finish 3`] = `
Array [
  Object {
    "FromInitOneCmd": true,
  },
  TaskMeta {
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  Array [
    Object {
      "b_1": ChildLogic {
        "args": Array [],
        "logicClass": [Function],
      },
      "b_2": ChildLogic {
        "args": Array [],
        "logicClass": [Function],
      },
    },
    [Function],
    Object {
      "FromInitTwoCmd": true,
    },
  ],
  Object {
    "b_1": ChildLogic {
      "args": Array [],
      "logicClass": [Function],
    },
    "b_2": ChildLogic {
      "args": Array [],
      "logicClass": [Function],
    },
  },
  Array [
    Object {
      "sharedDump": "{\\"some\\":{\\"FromInitOneCmd\\":true}}",
    },
    [Function],
  ],
  Object {
    "sharedDump": "{\\"some\\":{\\"FromInitOneCmd\\":true}}",
  },
  Array [
    Object {
      "FromInitOneCmd": true,
    },
    Event {
      "0": 3,
      "1": 2,
      "2": 1,
    },
  ],
  Object {
    "FromInitOneCmd": true,
  },
  Event {
    "0": 3,
    "1": 2,
    "2": 1,
  },
  Object {
    "ReactOnUpdateShared": Object {
      "0": 3,
      "1": 2,
      "2": 1,
    },
  },
  Object {
    "SharedUpdated": Object {
      "0": 3,
      "1": 2,
      "2": 1,
    },
  },
  Array [
    Object {
      "sharedDump": "{\\"some\\":{\\"FromInitOneCmd\\":true,\\"SharedUpdated\\":{\\"0\\":3,\\"1\\":2,\\"2\\":1}}}",
    },
    [Function],
  ],
  Object {
    "sharedDump": "{\\"some\\":{\\"FromInitOneCmd\\":true,\\"SharedUpdated\\":{\\"0\\":3,\\"1\\":2,\\"2\\":1}}}",
  },
  Array [
    Object {
      "FromInitOneCmd": true,
    },
    Event {
      "0": 3,
      "1": 2,
      "2": 1,
    },
  ],
  Object {
    "FromInitOneCmd": true,
  },
  Event {
    "0": 3,
    "1": 2,
    "2": 1,
  },
  Object {
    "ReactOnUpdateShared": Object {
      "0": 3,
      "1": 2,
      "2": 1,
    },
  },
  Object {
    "SharedUpdated": Object {
      "0": 3,
      "1": 2,
      "2": 1,
    },
  },
  Object {
    "FromInitOneCmd": true,
  },
  Object {
    "FromInitTwoCmd": true,
  },
  Array [
    Object {
      "b_1": ChildLogic {
        "args": undefined,
        "logicClass": [Function],
        "updateEvent": Event {
          "0": 1,
          "1": 2,
          "2": 3,
        },
      },
    },
    Object {
      "FromSubCmd": true,
    },
  ],
  Object {
    "b_1": ChildLogic {
      "args": undefined,
      "logicClass": [Function],
      "updateEvent": Event {
        "0": 1,
        "1": 2,
        "2": 3,
      },
    },
  },
  Object {
    "SomeEventUpdated": Object {
      "0": 1,
      "1": 2,
      "2": 3,
    },
  },
  Object {
    "FromSubCmd": true,
  },
  Array [
    [Function],
    Object {
      "FromPortCmd_2": true,
    },
  ],
  Object {
    "FromPortCmd_1": true,
  },
  Object {
    "FromPortCmd_2": true,
  },
  TaskMeta {
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  Object {
    "FromPortAsync_Success": true,
  },
  Object {
    "FromPortAsync_Success": true,
  },
]
`;
